verilog/analysis/checkers/dff_name_style_rule.cc

```diff
@@ -0,0 +1,288 @@
+          non_blocking_assigment,
+          absl::StrCat("Both parts before the suffix should be equal, but \"",
+                       lhs_base, "\" != \"", rhs_base, "\""),
+          context));
+    }
+  }
+}
+
+absl::string_view DffNameStyleRule::CheckSuffix(
+    const verible::SyntaxTreeContext &context, const verible::Symbol &root,
+    absl::string_view id, const std::vector<std::string> &suffixes) {
+  // Identifier is split between base and suffix:
+  // "myid_q" => {"myid", "_q"}
+  absl::string_view base;
+  // If there are no patterns to check against, everything passes the check
+  if (suffixes.empty()) return base;
+
+  // Lhs and Rhs should be plain variable identifiers
+  const verible::SyntaxTreeLeaf *leftmost_leaf = verible::GetLeftmostLeaf(root);
+  const verible::SyntaxTreeLeaf *rightmost_leaf =
+      verible::GetRightmostLeaf(root);
+
+  // a[2], a.b, ..., are not allowed
+  // Only simple identifiers and exceptions
+  if (leftmost_leaf != rightmost_leaf ||
+      leftmost_leaf->get().token_enum() != SymbolIdentifier) {
+    violations_.insert(LintViolation(
+        root,
+        absl::StrCat(
+            id, " Should be a simple reference, ending with a valid suffix: {",
+            absl::StrJoin(suffixes, ","), "}"),
+        context));
+
+    return base;
+  }
+
+  absl::string_view suffix_match;
+  // Check if id conforms to any valid suffix
+  const bool id_ok = std::any_of(
+      suffixes.begin(), suffixes.end(), [&](const std::string &suffix) {
+        if (absl::EndsWith(id, suffix)) {
+          base = absl::string_view(id.begin(), id.size() - suffix.size());
+          suffix_match = suffix;
+          return true;
+        }
+        return false;
+      });
+
+  // Exact match: id: "_q", suffix: "_q", there is no base
+  bool exact_match = suffix_match.size() == id.size();
+  if (id_ok && !exact_match) return base;
+
+  // The identifier can't be just the suffix we're matching against
+  const std::string lint_message =
+      exact_match
+          ? absl::StrCat(
+                "A valid identifier should not exactly match a valid prefix \"",
+                id, "\" == \"", suffix_match, "\"")
+          : absl::StrCat(id, " should end with a valid suffix: {",
+                         absl::StrJoin(suffixes, ","), "}");
+  violations_.insert(LintViolation(root, lint_message, context));
+
+  return base;
+}
+
+absl::Status DffNameStyleRule::Configure(absl::string_view configuration) {
+  // If configuration is empty, stick to the default
+  if (configuration.empty()) return absl::OkStatus();
+
+  std::string output, input;
+  absl::Status status = verible::ParseNameValues(
+      configuration,
+      {{"output", SetString(&output)}, {"input", SetString(&input)}});
+
+  if (!status.ok()) return status;
+
+  valid_input_suffixes = ProcessSuffixes(input);
+  valid_output_suffixes = ProcessSuffixes(output);
+
+  return status;
+}
+
+std::vector<std::string> DffNameStyleRule::ProcessSuffixes(
```

Comment:
Could this return a `std::vector<std::string_view>` instead?  I believe the configuration objects lifetime makes this safe.

Reply:
The result vector will contain modified config elements though, so we can't return them as string-views coming from the original configuration.

But what will work to avoid having to copy into `std::string` twice is to have the original split be returning string-views, but then generate std::strings in the transform:

```c++
const std::vector<absl::string_view>  split_suffixes = absl::StrSplit(config, ',', absl::SkipEmpty());

std::vector<std::string> result(split_suffixes.size());
std::transform(split_suffixes.cbegin(),..., result.begin(),...);

return result;
```

Reply:
Solved, thanks for the catch.